### ISSUE TITLE
Throubles with access to POP3 mailbox using Python v3

### ISSUE THREAD

**Issue by COshmyan** [jorritfolmer/TA-dmarc#34]
Splunk Version 8.1.4
TA-dmarc Version 3.2.4

It seems that this add-on does not work with POP3 server using Python v3 (default Python version since Splunk v8.x).
The previous issues (#31 and #32) mentioned that this add-on shoud work with Python v3 also; however, I did not find any notes specific for this combination (Python3 + POP3 mailbox).
The typical error messages in the log file are the following:

```
2021-05-19 10:28:00,234 INFO pid=24336 tid=MainThread file=setup_util.py:log_info:117 | Log level is not set, use default INFO
2021-05-19 10:28:00,234 INFO pid=24336 tid=MainThread file=setup_util.py:log_info:117 | Log level is not set, use default INFO
2021-05-19 10:28:00,234 INFO pid=24336 tid=MainThread file=base_modinput.py:log_info:295 | Start processing pop3 server ldnode2.rietumu.lv with use_ssl True
2021-05-19 10:28:00,385 INFO pid=24336 tid=MainThread file=base_modinput.py:log_info:295 | get_dmarc_messages: 79 messages
2021-05-19 10:28:00,386 INFO pid=24336 tid=MainThread file=splunk_rest_client.py:_request_handler:105 | Use HTTP connection pooling
2021-05-19 10:28:00,832 INFO pid=24336 tid=MainThread file=base_modinput.py:log_info:295 | Start processing 79 new messages of 79 on ldnode2.rietumu.lv
2021-05-19 10:28:00,834 ERROR pid=24336 tid=MainThread file=base_modinput.py:log_error:309 | Get error when collecting events.
Traceback (most recent call last):
  File "/opt/splunk/etc/apps/TA-dmarc/bin/ta_dmarc/aob_py3/modinput_wrapper/base_modinput.py", line 128, in stream_events
    self.collect_events(ew)
  File "/opt/splunk/etc/apps/TA-dmarc/bin/dmarc_pop3.py", line 84, in collect_events
    input_module.collect_events(self, ew)
  File "/opt/splunk/etc/apps/TA-dmarc/bin/input_module_dmarc_pop3.py", line 46, in collect_events
    filelist = p2d.process_incoming()
  File "/opt/splunk/etc/apps/TA-dmarc/bin/dmarc/pop2dir.py", line 281, in process_incoming
    response = self.get_dmarc_message_bodies(new_messages)
  File "/opt/splunk/etc/apps/TA-dmarc/bin/dmarc/pop2dir.py", line 105, in get_dmarc_message_bodies
    msg = "\n".join(self.server.retr(uid.split()[0])[1])
  File "/opt/splunk/lib/python3.7/poplib.py", line 248, in retr
    return self._longcmd('RETR %s' % which)
  File "/opt/splunk/lib/python3.7/poplib.py", line 183, in _longcmd
    return self._getlongresp()
  File "/opt/splunk/lib/python3.7/poplib.py", line 159, in _getlongresp
    resp = self._getresp()
  File "/opt/splunk/lib/python3.7/poplib.py", line 152, in _getresp
    raise error_proto(resp)
poplib.error_proto: b'-ERR No message 0. Maildrop contains 79 message(s).'
```

(the more detailed output with level=DEBUG is attached).

Note the following:

- In Python 2 strings (the "str" type) are a really byte strings; but in Python 3 strings are Unicode strings and just a byte strings have a separate type - "bytes" (see details [here](https://sebastianraschka.com/Articles/2014_python_2_3_key_diff.html#unicode));
- It is possible to convert a byte strings into Unicode strings and vise versa (using a specific encoding), but it's impossible just to concatenate a strings of different types (see the last example by the link above);
- In fact, the [poplib](https://docs.python.org/3/library/poplib.html) module (even in Python v3) works with octets (i.e. byte strings): for example, the **uidl()** method returns not a Unicode string but a byte string (it is obviously in log with DEBUG level);
- The **retr()** method accepts either Unicode string or Numeric (Integer) as a parameter. However, add-on sends the byte string here; so, it causes to error: the byte string is not converted onto Unicode string correctly, so the POP3-server response with the "-ERR No message 0." error. The message with number 0 (zero) is really does not exist, moreover, according to [POP3 protocol](https://datatracker.ietf.org/doc/html/rfc1939#page-4) is a positive Integer:

> The first message in the maildrop is assigned a message-number of
   "1", the second is assigned "2", and so on, so that the nth message
   in a maildrop is assigned a message-number of "n".

Probably, some explicit conversion between byte-strings and Unicode-strings are needed (using, for example, UTF-8 encoding) for POP3 access.

I tried to use a Python2 as a workaround; but, unfortunately, without success (according to logs, the Python3 is still used). I did set the following section:
`[general]
python.version = python2
`
in both "server.conf" and "inputs.conf" config files in "apps/TA-dmarc/local" directory; but anyway it seems that add-on uses the Python3.
[ta_dmarc_dmarc_pop3.log](https://github.com/jorritfolmer/TA-dmarc/files/6532826/ta_dmarc_dmarc_pop3.log)

**Comment by jorritfolmer** [comment]
Thanks for posting this issue, I'll look into it in a couple of weeks.

**Comment by jorritfolmer** [comment]
It seems you have a pretty good idea what causes the issue here, so I'm more than happy to accept a pull request that fixes it.

**Comment by COshmyan** [comment]
Hi @jorritfolmer,
Thank you for a fast reply.

Unfortunately, I'm not a Python programmer, so I could not propose an appropriate patch.
However, I'm able to share my analysis; probably, it will help.

- First of all, it seems that problem is in file "TA-dmarc/bin/dmarc/pop2dir.py".
- This file has the following fragment in **get_dmarc_message_bodies()** method:

>         for uid in messages:
>             self.helper.log_debug('get_dmarc_message_bodies: got uid "%s", using uid "%s"' % (uid, uid.split()[0]))
>             msg = "\n".join(self.server.retr(uid.split()[0])[1])
>             msgobj = email.message_from_string(msg)
>             if msgobj.get("Subject").find("Report domain:") >= 0:
>                 self.helper.log_debug(
>                     'get_dmarc_message_bodies: found dmarc message: uid %s with subject %s' %
>                     (uid, msgobj.get("Subject")))
>                 response[uid] = {}
>                 response[uid][b'RFC822'] = msg
>         return response

- The log_debug() method here 
>self.helper.log_debug('get_dmarc_message_bodies: got uid "%s", using uid "%s"' % (uid, uid.split()[0]))

is still working, but the next line (with a call of "**self.server.retr(uid.split()[0])[1]**" method) is already fails.
- even this **log_debug()** call displays that it receives a byte string (pay attension to the "**b'**" prefix for both values:

> get_dmarc_message_bodies: got uid "b'6 81628DD8C842D229428B882CD8B9C5B3'", using uid "b'6'"

- The "messages" (processed in **get_dmarc_message_bodies()** method) is the array obtained by the **get_dmarc_messages()** method (see the last line here):

>            messages = self.server.uidl()[1]

- My intuition tells me that this is the best place for conversion from bytes string to Unicode string (using, for example, the **decode('utf-8')** method); however, as I'm not a Python programmer, I could not offer: how it should be performed properly with an array;
- Probably, there are another places also during a processing of messages (for example, analyzing a Subject of message).

Hope, it helps.

**Comment by COshmyan** [comment]
I'm not sure if somebody still uses Python v2, but ideally solution should continue to work in this version, too.
So, it's needed some check (if the transformation is needed or not).
Something like the following (in the **get_dmarc_messages()** method):

>messages = self.server.uidl()[1]
>\+if len(message)>0 and type(message[0]) == "bytes"
>\+  messages = [ s.decode("utf-8") for s in messages ]

At the same time, I'm still not sure if it is enough.

**Comment by COshmyan** [comment]
Hi @jorritfolmer,

I've produced a pull request [(link)](https://github.com/jorritfolmer/TA-dmarc/pull/35), please, review it.
It was successfully tested in our environment (Splunk v8.1.4, Python v3, TA-dmarc v3.2.4).
Probably, should work with Python v2 too, but was not tested in this environment.

This patch also contains one more bug fix.
Th old version processed only e-mails that contained string "Report domain:" in Subject (using a case-sensitive search):

`if msgobj.get("Subject").find("Report domain:") >= 0:`

However, in our practice only google.com sends reports with exactly this Subject.
All other use the string: "Report Domain:" (where "Domain" word starts with uppercase letter).
So, it's better to use a case-insensitive comparison.
Additionally, an [official Python manual](https://docs.python.org/3/library/stdtypes.html#text-sequence-type-str) tells:

> **Note** The **find()** method should be used only if you need to know the position of _sub_. To check if _sub_ is a substring or not, use the **in** operator:
`>>> 'Py' in 'Python'`
`True`

So, I've rewrote this line in the following manner:

`if "report domain:" in msgobj.get("Subject").lower():`

It works better, processing all necessary e-mails.
--
Sincerely yours,
Constantin Oshmyan

**Comment by jorritfolmer** [comment]
Thanks for your contribution! I’ll look into it this weekend!

**Comment by jorritfolmer** [comment]
Merged!

**Comment by jorritfolmer** [comment]
I'll release v3.2.5 with your fixes later because there are some issues on Splunkbase. I'll leave this open for now and close once released.

**Comment by jorritfolmer** [comment]
Released 4.0.0 several weeks ago on Splunkbase that includes as well.
